### PR TITLE
[SPARK-39759][SQL] Implement listIndexes in JDBC (H2 dialect)

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Implementing listIndexes in DS V2 JDBC for H2 dialect.

### Why are the changes needed?
This is a subtask of the V2 Index support(https://issues.apache.org/jira/browse/SPARK-36525). 
**It can better test the index interface locally.**
> This PR implements listIndexes in H2 dialect.

### Does this PR introduce _any_ user-facing change?
Yes, listIndexes in DS V2 JDBC for H2

### How was this patch tested?
Update existed UT.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 1]
```diff
@@ -110,6 +112,64 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+          val indexComment = rs.getString("REMARKS")
+          if (indexMap.contains(indexName)) {
+            val index = indexMap(indexName)
+            val newIndex = new TableIndex(indexName, indexType,
+              index.columns() :+ FieldReference(colName),
+              index.columnProperties, index.properties)
+            indexMap += (indexName -> newIndex)
+          } else {
+            val properties = new util.Properties()
+            if (StringUtils.isNotEmpty(indexComment)) properties.put("COMMENT", indexComment)
```
Comment (panbingkun) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:158]: In fact, I found possible null pointer in the implementation of listIndexes(H2), indexComment default value is null.
However, the default value of MySQLDIalect is empty String.
@huaxingao

**Comment by panbingkun** [comment]
ping @huaxingao

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 2]
```diff
@@ -110,6 +112,64 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+  // https://www.h2database.com/html/systemtables.html?#information_schema_indexes
+  // https://www.h2database.com/html/systemtables.html?#information_schema_index_columns
+  override def listIndexes(
+      conn: Connection,
+      tableIdent: Identifier,
+      options: JDBCOptions): Array[TableIndex] = {
+    val sql = {
+      s"""
+         | SELECT
+         |  i.INDEX_CATALOG AS INDEX_CATALOG,
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:125]: nit: I think the indentation should be two spaces. 
```
         | SELECT
         |   i.INDEX_CATALOG AS INDEX_CATALOG,
```

**Comment by huaxingao** [comment]
Does H2 support composite index?

**Review by huaxingao — APPROVED** [review]

**Comment by huaxingao** [comment]
Merged to master. Thanks @panbingkun

**Pull request closed without merging**